### ISSUE TITLE
Deep MNIST for Experts Tutorial accuracy is lower than expected (0.9091)

### ISSUE THREAD

**Issue by connormclaud** [tensorflow/tensorflow#2781]
### Environment info
Operating System:

> $ uname -a
> Linux vvinahradski 4.4.0-23-generic #41-Ubuntu SMP Mon May 16 23:04:25 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
Ubuntu 16.04

Installed version of CUDA and cuDNN: 
(please attach the output of `ls -l /path/to/cuda/lib/libcud*`):
Not installed

If installed from binary pip package, provide:

1. Which pip package you installed.
https://storage.googleapis.com/tensorflow/linux/cpu/tensorflow-0.9.0rc0-cp35-cp35m-linux_x86_64.whl
2. The output from `python -c "import tensorflow; print(tensorflow.__version__)"`.
0.9.0rc0

### Steps to reproduce
1. Copy/past https://www.tensorflow.org/versions/r0.9/tutorials/mnist/pros/index.html#deep-mnist-for-experts tutorial
2. Run code
3. Observe that accuracy is around 90% instead of expected 99%

### What have you tried?
1. Increased test dataset 10x times. Accuracy is around 94%

### Logs or other output that would be helpful
(If logs are large, please upload as attachment).
Source code:
[tut.zip](https://github.com/tensorflow/tensorflow/files/308938/tut.zip)
console output:
[log.txt](https://github.com/tensorflow/tensorflow/files/308940/log.txt)

**Comment by girving** [comment]
@keveman Can you take a look?
@vrv, @martinwicke If real this could be another release blocker.

**Comment by vrv** [comment]
If this doesn't happen in 0.8, then yes, it's a blocker.

**Comment by keveman** [comment]
@girving taking a look.

**Comment by kingtaurus** [comment]
Ran a quick test:
```
python --version
Python 3.4.3+
```

Relevant details:
```
python -c "import tensorflow; print(tensorflow.__version__)"
0.9.0rc0
```

Got the following:
```
step 19800, training accuracy 0.9
step 19900, training accuracy 0.88
test accuracy 0.9071
```

Changing the learning rate to (1e-3):
```
step 19800, training accuracy 0.96
step 19900, training accuracy 1
test accuracy 0.9623
```

Which looks a little better.

**Comment by keveman** [comment]
0.8.0 and 0.9.0rc0 get the same accuracy (modulo differences due to random initialization). @kingtaurus thanks for checking the learning rate.

**Comment by girving** [comment]
@keveman What is that same accuracy?  This tutorial is supposed to get to 99.2% accuracy.  If it's the same model as `models/image/mnist/convolutional.py`, that's what it should be getting.

**Comment by yaroslavvb** [comment]
The models look identical except Deep MNIST for experts uses 1024
<https://www.tensorflow.org/versions/r0.9/tutorials/mnist/pros/index.html#densely-connected-layer>
for first FC layer and Vincent's model uses 512
<https://github.com/tensorflow/tensorflow/blob/6431560b7ec3565154cb9cdc9c827db78ccfebe7/tensorflow/models/image/mnist/convolutional.py#L174>

Also, Deep MNIST uses Adam, whereas convolutional.py uses momentum
with staircase
learning rate decay
<https://github.com/tensorflow/tensorflow/blob/6431560b7ec3565154cb9cdc9c827db78ccfebe7/tensorflow/models/image/mnist/convolutional.py#L245>

**Comment by kingtaurus** [comment]
Its not the same model:
(1) this model uses `tf.train.AdamOptimize(1e-4)` instead of (`tf.train.exponential_decay` combined with `tf.train.MomentumOptimizer` and starts at a learning rate of `0.01`)
(2) 
```
y_conv=tf.nn.softmax(tf.matmul(h_fc1_drop, W_fc2) + b_fc2)
tf.reduce_mean(
        -tf.reduce_sum(y_ * tf.log(y_conv), reduction_indices=[1]))
```
is used for loss (rather than `tf.nn.sparse_softmax_cross_entropy_with_logits`) [this should be identical but just a minor point] ;
(3) regularization is being applied in (`models/image/mnist/convolutional.py`);
(4) `convolutional.py` does 10 epochs of batch size 64;
     `tut.py` does 18 epochs of batch size 64;
(5) Finally the different size of the Full connected layers;

**Comment by girving** [comment]
@vincentvanhoucke Should the mnist for experts tutorial be the same model as `models/image/mnist/convolutional.py`?

@keveman When you say the accuracy is the same, is it close to 99.2%?

**Comment by keveman** [comment]
@girving No, both were close to 90%

**Comment by girving** [comment]
@keveman We should probably fix the tutorial to be 99% then, unless there is a compelling reason to make it 10x worse than `models/image/mnist/convolutional.py`.

**Comment by connormclaud** [comment]
In my humble opinion, this issue may indicate regression in core nn code.  I believe tutorial was correct at some point.

**Comment by kingtaurus** [comment]
There is **bug** in your code. You are only running the learning step once every 100 batches;

```
        for i in range(20000):
            batch = mnist.train.next_batch(50)
            if i % 100 == 0:
                train_accuracy = accuracy.eval(
                    feed_dict={
                        x:batch[0], y_: batch[1], keep_prob: 1.0})
                print("step %d, training accuracy %g"%(i, train_accuracy))
                train_step.run(
                    feed_dict={x: batch[0], y_: batch[1], keep_prob: 0.5})
```

Should be:
```
        for i in range(20000):
            batch = mnist.train.next_batch(50)
            if i % 100 == 0:
                train_accuracy = accuracy.eval(
                    feed_dict={
                        x:batch[0], y_: batch[1], keep_prob: 1.0})
                print("step %d, training accuracy %g"%(i, train_accuracy))
            train_step.run(
                    feed_dict={x: batch[0], y_: batch[1], keep_prob: 0.5})
```

**Comment by keveman** [comment]
Good catch @kingtaurus. I can confirm ~99% accuracy with 0.9.

**Comment by basuam** [comment]
@kingtaurus

Thank you for catching up that indentation mistake.

When I corrected it, I got the same accuracy they said on the tutorial:  **0.9922**

step 19000, training accuracy 1
step 19100, training accuracy 1
step 19200, training accuracy 1
step 19300, training accuracy 1
step 19400, training accuracy 1
step 19500, training accuracy 1
step 19600, training accuracy 1
step 19700, training accuracy 1
step 19800, training accuracy 1
step 19900, training accuracy 1
test accuracy 0.9922
